cpp/src/umap/simpl_set_embed/algo.h

```diff
@@ -77,16 +77,15 @@ void make_epochs_per_sample(T *weights, int weights_n, int n_epochs, T *result,
       if (v * n_epochs > 0)
         return T(n_epochs) / v;
       else
-        return T(-1);
+        return -1.0f;
     },
     stream);
 }
 
 /**
-	         * Clip a value to within a lower and upper bound
-	         */
-template <typename T>
-__device__ __host__ T clip(T val, T lb, T ub) {
+ * Clip a value to within a lower and upper bound
+ */
+__device__ __host__ double clip(double val, double lb, double ub) {
```

Comment:
Same here. No need for custom kernel functions. You can use 

https://github.com/rapidsai/cuml/blob/branch-0.10/cpp/src_prims/linalg/binary_op.h

There are examples of these types of functions in cuML that uses prims rather than custom kernels.

Reply:
This isn’t a custom kernel, it’s just a __device__ function for clipping a value and I didn’t see it reusable enough to muddy up our prims with it. It’s used by an actual kernel  to clip the gradient.